migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql

```diff
@@ -0,0 +1,8 @@
+ALTER TABLE webhooks
+    DROP COLUMN IF EXISTS created_by,
+    DROP COLUMN IF EXISTS updated_by,
+    ADD COLUMN created_by INTEGER,
+    ADD COLUMN updated_by INTEGER;
```

Comment:
I kinda feel strongly about this 😄 Matches also a lot of other foreign keys we have: https://sourcegraph.sourcegraph.com/search?q=file%3Aschema.md+REFERENCES+users%28id%29&groupBy=repo

```suggestion
    ADD COLUMN created_by_user_id INTEGER,
    ADD COLUMN updated_by_user_id INTEGER;
```

I'd also make this a proper foreign key.

Reply:
I agree with a name change so it is more clear.

But we don't want to make it a foreign key, because user deletion should not affect the webhook.

Reply:
@mrnugget We're going to experiment with making the columns nullable and adding the foreign key constraint. Just want to test that when the user is deleted the column is set to null.. at least that's what we'd expect.

Reply:
You can have foreign keys without `ON DELETE CASCADE`. You can have it set to `NULL`, yes.

Reply:
that's the plan. now the only thing is to put it into golang's int32 so it doesn't look ugly

I think about mapping `NULL` from the DB to `-1` in `types.Webhook`. Cannot really think about other way of doing that without changing the type

Reply:
but the problem is that we indeed need a special user with id==0 to flag that the webhook is created programmatically...

Reply:
So you want to have 3 states? user exists, user deleted (_do_ we hard delete users? I thought we'd only soft-delete them), user is "machine". I'd add a `auto_created boolean`, to be honest. And have `UserID int` in the `types.Webhook` type. Then you can say `if !webhook.AutoCreated && webhook.UserID == 0 { // user hard-deleted }`.

Reply:
### Three states

Yep, I thought about the same thing.
This is not the most pretty-looking solution, especially because we would need both `auto_created` and `auto_updated` columns, but seems the only one to support 3 states and have a foreign key to `users` table at the same time.

### User deletion

Also, you've mentioned soft and hard deleting. We actually do both, but hard-deleting a user is only performed from site admin screen.
I took a look at how users are deleted (both [soft](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@202a6e7d89ef5bc7b85638fbbfd2510c236974ea/-/blob/internal/database/users.go?L513:21) and [hard](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@202a6e7d89ef5bc7b85638fbbfd2510c236974ea/-/blob/internal/database/users.go?L570)) and although we have a bunch of foreign keys from other tables, deleting a user involves a bunch of queries ran across multiple tables in one transaction.

### Which gave me another option to implement what we want

Don't add a foreign key and keep it in one column (-1 for deleted user, 0 for auto_created and >0 for user ID) and add another query to clean up `webhooks` table when the user is deleted.

WDYT?
@ryanslade @mrnugget

Reply:
> Don't add a foreign key and keep it in one column (-1 for deleted user, 0 for auto_created and >0 for user ID) and add another query to clean up webhooks table when the user is deleted.

I actually feel strongly about this and that we shouldn't do it. [In _a lot_ of parts of our codebase `0` on a `*UserID` field means either "user doesn't exist" or "doesn't exist anymore" or "no user was set".](https://sourcegraph.sourcegraph.com/search?q=UserID+%3D%3D+0&groupBy=repo). That also makes it easy to handle `NULL` in the database: it'll be a `0` in the Go type. So on hard-delete of a user you can set the column to `NULL` and you get `0` in Go and you don't need to wrestle with pointers.

I'd add a second `boolean` column _once_ we have auto-creation of webhooks (that's part of milestone 4, right? https://docs.google.com/document/d/1cldGAS0fz3wqa_9Ny1zv6SHCsIrB5EsljRWPFXXoRAg/edit#heading=h.9lqrs4lkaqow))

Reply:
That's nice to know about 0 user ID being "mapped" to "no user", we should go this way here as well.

Then what you proposed looks like the best option, I will implement it this way.

Adding an `auto_created` column can be postponed until the time we really need it, I agree 👍🏻

Reply:
Awesome, sounds good to me 👍